bootstrap.md

```diff
@@ -0,0 +1,136 @@
+
+At its simplest, bootstrapping can be thought of as secure network joining. It must enable a new node to get the latest head/chain that the majority of the network power is mining off and must be both:
+
+- Robust — able to do so in adversarial networks (so long as honest nodes make up the majority of the network).
+
+- Consistent — able to reliably get the canonical chain (even given poor latency or state loss in our initial peer set).
+
+The specific bootstrapping heuristic could be left up to the node implementation, but we provide a simple one here as a candidate construction. This spec is divided as follows:
+
+- Initial Connection — ensuring the node connects to an honest set of initial peers.
+- Peer Set Expansion — ensuring that these peers can serve up the network's canonical head reliably in spite of their individual latencies or state loss.
+- Chain Selection — choosing the most likely canonical chain thereafter.
+
+We leave trustless initial peer selection as future work.
+
+## Initial Connection
+
+Initial peer selection is key to avoiding eclipse attacks. If the nodes you initially connect to are malicious, then they can simply serve you malicious other peers to all of your requests, and effectively prevent you from ever discovering the real network. Filecoin assumes 51% of rational or honest miners. Given no prior information, as is the case when bootstrapping, distinguishing a rational node from an adversarial one would require sampling a majority of the network.
+
+This is obviously impractical, so we choose to start the bootstrap sequence with nodes by connecting with an inital set of trusted peers. These can be nodes run by well-known entities (the Filecoin foundation, large exchanges, good actors in the community, etc) or manually specified by the node operator in the config file. We leave trustless (or fully decentralized) bootstrapping for future work.
+
+A strawman bootstrapping sequence would have nodes use these initial peers to sync up to the canonical chain. However, new connection requests might effecitvely DoS these peers, making the Filecoin network unavailable to new nodes. Therefore, Filecoin bootstrapping seeks to "load-balance" initial requests to ensure the network remains widely available.
+
+We connect to each of these initial peers using the `hello handshake` (see in [network protocols](./network-protocols.md#hello-handshake)), getting back
+
+```go
+type HelloMessage struct {
+HeaviestTipSet []Cid
+HeaviestTipSetHeight uint64
+GenesisHash Cid
+}
+```
+
+followed by a `GetPeers` request, getting back
```

Comment:
Yes as you suspect right now we do not receive our peers' peers by querying directly like this.  Instead [we](https://github.com/filecoin-project/go-filecoin/blob/master/net/bootstrap.go#L111) do [kademlia DHT bootstrapping](https://github.com/libp2p/go-libp2p-kad-dht/blob/master/dht_bootstrap.go#L149).

Reply:
this is an interesting thought. Are the peers kademlia is returning the 'proper' set of peers? 

If so we can continue using it, if not we may need a separate protocol.

If we keep using kademlia for this, we should expose the single 'getPeers' request (kademlia `FIND_NODE` or whatever) so that we can use it here. 

Alternatively, we should ensure that kademlia bootstrapping follows this algorithm exactly.

Reply:
I'm unsure what peers you're trying to get? You want the peers of the peer you just connected to, but with respect to what? In DHT you want some target, so you might get the contents of the closest bucket in the routing table. Or do you just want whomever the peer is connected to, in which case it's entirely unrelated to the DHT?

Reply:
It does feel like you're trying to get a random set of peers, in which case the only value the DHT is providing is that its routing table properties and basic operations should ensure a reasonably distributed interconnectivity of all participants. That is that by knowing one node in the DHT, one can traverse to any other node in the DHT.

Reply:
Yes, I'm unclear what you mean by "'proper' set of peers" @whyrusleeping, but I'm on the same page @anacrolix for all intents and purposes, I'd want to be able to get back n random peers to get some representative sample of the network calling this enough times.

Reply:
@sternhenri yeah, its not clear to me what exactly 'proper' means here. You tell me. All i'm saying is that if you provide a definition of 'proper' and the DHT satisfies that, then thats great! otherwise we will need to implement a specific protocol for this.

Reply:
After a conversation with @anacrolix, we have a strategy. First, to define *proper*, we mean a random peer. Basically, we want to query peers to make up a "representative" set of the network, without risking getting peers from given clusters. Specifically, we are looking for this set to be pretty random (and not tied to network topography, latency, etc) to
1) avoid hotspots
2) make eclipse harder to build

So we want to treat the Filecoin nodes and underlying libp2p network as two distinct networks for bootstrapping purposes. While we'll use the DHT later for routing, we don't allow it to affect choice of peers early on.

Typical DHT bootstrapping would mean querying your own ID (to get those nodes closest to you). Instead here, we will successively query random IDs (25 times) round-robinning our trusted nodes as starting points. We'll keep the nodes with IDs closest to our random one in our peer set.

We use libp2p's [GetClosestPeers](https://godoc.org/github.com/libp2p/go-libp2p-kad-dht#IpfsDHT.GetClosestPeers) fn.

All this implies Fil node IDs are randomly generated (so we don't accidentally get back a cluster).